File: wrappers/javascript/aries-askar-shared/src/store/Entry.ts
import type { EntryListHandle } from '../crypto'

import { Buffer } from 'buffer'

export type EntryObject = {
  name: string
  value: Record<string, unknown> | string
  tags: Record<string, unknown>
  category: string
}

export class Entry {
  private _list: EntryListHandle
  private _pos: number

  public constructor({ list, position }: { list: EntryListHandle; position: number }) {
    this._list = list
    this._pos = position
  }

  public get category() {
    return this._list.getCategory(this._pos)
  }

  public get name() {
    return this._list.getName(this._pos)
  }

  public get value(): string {
    return Buffer.from(this.rawValue).toString()
  }

  private get rawValue() {
    return this._list.getValue(this._pos)
  }

  public get tags() {
    const tags = this._list.getTags(this._pos)

    if (!tags) return {}
    return JSON.parse(tags) as Record<string, unknown>
  }

  public toJson(shouldParseValueToJson = false): EntryObject {
    return {
      name: this.name,
      value: shouldParseValueToJson ? (JSON.parse(this.value) as Record<string, unknown>) : this.value,
      tags: this.tags,
      category: this.category,
    }
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/Store/Entry.kt
package askar.Store

import askar.crypto.EntryListHandle
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

/***
 * @param category the string name of the category
 * @param name the string name of the entry
 * @param value the value passed in when the entry was created
 * @param tags a json formatted string of tags
 */
@Serializable
class EntryObject(val category: String, val name: String, val value: String, val tags: String) {

    override fun toString(): String {
        val temp = buildJsonObject {
            put("category", category)
            put("name", name)
            put("value", value)
            val tagsJson = Json.decodeFromString<JsonElement>(tags)
            put("tags", tagsJson)
        }
        return temp.toString()
    }

    override fun equals(other: Any?): Boolean {
        if(other == null) return false
        val o = other as EntryObject
        val tags = Json.decodeFromString<JsonObject>(this.tags)
        val otherTags = Json.decodeFromString<JsonObject>(o.tags)
        return o.category == this.category && o.name == this.name && o.value == this.value && tags == otherTags
    }

}
class Entry (private val list: EntryListHandle, private val pos: Int) {

    fun category(): String {
        return this.list.getCategory(this.pos)
    }

    fun name(): String {
        return this.list.getName(this.pos)
    }

    fun value(): String {
        return this.list.getValue(this.pos)
    }

    //Revisit if needed later
//    private fun rawValue(): String {
//        return this.list.getValue(this.pos)
//    }

    fun tags(): String {
        return list.getTags(pos) ?: "{}"
    }

    fun toJson(): EntryObject {
        val entry = EntryObject(category(), name(), tags = tags(), value = value())
       return entry
    }

}